File: take-gifts-from-the-richest-pile/take-gifts-from-the-richest-pile.cpp
class Solution {
public:
    long long pickGifts(vector<int>& gifts, int k)
    {
        long long len = gifts.size();
        long long res = 0;
        
        priority_queue<long long>que;
        
        for(auto i: gifts)
        {
            que.push(i);
        }
        
        while(k--)
        {
            que.push(sqrt(que.top()));
            que.pop();
        }
        
        while(!que.empty())
        {
            res += que.top();
            
            que.pop();
        }
        
        return res;
    }
};

File: 2692-take-gifts-from-the-richest-pile/take-gifts-from-the-richest-pile.py
class Solution:
    def pickGifts(self, gifts: List[int], k: int) -> int:
        heap = []

        for gift in gifts:
            heap.append(-gift)

        heapq.heapify(heap)

        while k:
            top = -heapq.heappop(heap)
            sqt = int(sqrt(top))
            heapq.heappush(heap, -sqt)
            k -= 1

        ans = 0

        for gift in heap:
            ans += (-1 * gift)

        return ans